### PR TITLE
[#174416949] ESLint config

### PR DESCRIPTION
**Pull request by BurnedMarshal**
Porting TSLint rules configuration from `io-tslint-rules` to ESLint and merged with `io-pay` ESLint configuration .

The following TSLint rules are not supported for eslint at the moment:
```
bool-param-default
max-union-size
no-accessor-field-mismatch
no-array-delete // Mitigated by no-delete
no-case-with-or
no-dead-store
no-duplicate-in-composite
no-empty-array
no-extra-semicolon // Mitigated by prettier
no-empty-destructuring // Mititgated by no-empty-pattern
no-gratuitous-expressions
no-hardcoded-credentials
no-ignored-initial-value // Mitigated by no-param-reassign, no-let
no-in-misuse
no-inferred-empty-object-type
no-invalid-await // Mititgated by await-thenable
no-invariant-return
no-misleading-array-reverse // Mitigated by immutable-data
no-misspelled-operator // Mitigated by prettier
no-nested-switch
no-nested-template-literals
no-statements-same-line // Mitigated by prettier
no-try-promise
no-tslint-disable-all
no-unconditional-jump
no-undefined-argument
no-unenclosed-multiline-block // Mitigated by prettier
no-unthrown-error
no-unused-array // Mitigated by no-unused-vars
no-useless-increment
no-useless-intersection
prefer-promise-shorthand
promise-must-complete
use-primitive-type
```

### CODE REVIEW

File: CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1 @@
+@BurnedMarshal @AleDore @gunzip @cloudify @francescopersico
```
Comment (gunzip) [inline CODEOWNERS]:
```suggestion
* @BurnedMarshal @AleDore @gunzip @cloudify @francescopersico
```

File: package.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "io-eslint-rules",
```
Comment (gunzip) [inline package.json]: why the package and repo names diverge?

File: recommended.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    "@typescript-eslint/prefer-function-type": "error",
+    "@typescript-eslint/restrict-plus-operands": "error",
+    "semi": "off",
+    "@typescript-eslint/semi": ["error"],
+    "@typescript-eslint/unified-signatures": "error",
+    "functional/no-let": "error",
+    "functional/immutable-data": "error",
+    "sonarjs/no-small-switch": "off",
+    "sonarjs/no-duplicate-string": "error",
+    "sonarjs/cognitive-complexity": ["error", 19],
```
Comment (AleDore) [inline recommended.js:85]: We must check on TSLint what is the current cap on cognitive complexity

**Review by AleDore — COMMENTED** [review]
LGTM apart from minor tuning

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,70 @@
+{
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig.json to read more about this file */
+
+    /* Basic Options */
+    // "incremental": true,                   /* Enable incremental compilation */
+    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
```
Comment (gunzip) [inline tsconfig.json]: why not es6?

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
+    "lib": ["es2015"],                             /* Specify library files to be included in the compilation. */
```
Comment (gunzip) [inline tsconfig.json]: es6?

File: tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,70 @@
+{
```
Comment (gunzip) [inline tsconfig.json]: why do we need a tsconfig.json here?

**Review by gunzip — COMMENTED** [review]
can you add a readme.md with the user manual ? (how to integrate in new projects and how to migrate from tslint in existing ones)

File: recommended.js (hunk 3 above)
Comment (BurnedMarshal) [reply recommended.js:85]: The default value is `15`, we can use it and eventually tuning the value at the project level.

File: tsconfig.json (hunk 6 above)
Comment (BurnedMarshal) [reply tsconfig.json]: It was added for local testing inside this project. We can safely remove it.  Fixed in 1a44979

File: tsconfig.json (hunk 4 above)
Comment (BurnedMarshal) [reply tsconfig.json]: See https://github.com/pagopa/eslint-rules/pull/1#discussion_r546698816

File: tsconfig.json (hunk 5 above)
Comment (BurnedMarshal) [reply tsconfig.json]: See https://github.com/pagopa/eslint-rules/pull/1#discussion_r546698816

File: package.json (hunk 2 above)
Comment (BurnedMarshal) [reply package.json]: The name of the package must be `@pagopa/eslint-config` to work properly. See [this](https://eslint.org/docs/developer-guide/shareable-configs#sharing-multiple-configs) ESLint documentation reference. If we want we can rename the repo to `eslint-config`. Fixed in 7dd1a75

**Comment by BurnedMarshal** [comment]
> can you add a readme.md with the user manual ? (how to integrate in new projects and how to migrate from tslint in existing ones)

@gunzip A small README content was added in 32366af and 96e3e9e

**Review by gunzip — APPROVED** [review]

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1,6 @@
+node_modules
+yarn-error.log
+tsconfig.json
+src/**
+.eslintrc.js
+.eslintignore
```
Comment (balanza) [inline .gitignore]: End file with new line (there's aN ESLINT rule for that as well ;) )
```suggestion
.eslintignore

```

File: .prettierrc.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+  trailingComma: 'none',
+  arrowParens: 'avoid',
+  semi: true,
+  trailingComma: 'all',
+  singleQuote: false,
+  printWidth: 120,
+  tabWidth: 2,
+};
```
Comment (balanza) [inline .prettierrc.js]:
```suggestion
};

```

File: recommended.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    "no-inner-declarations": "off",
+    "prefer-const": "error",
+    curly: "error",
+    "spaced-comment": ["error", "always", { block: { balanced: true } }],
+    radix: "error",
+    "one-var": ["error", "never"],
+    "object-shorthand": "error",
+    "no-var": "error",
+    "no-param-reassign": "error",
+    "no-underscore-dangle": "error",
```
Comment (balanza) [inline recommended.js:30]: I'm not sure if it would break our convention of prefixing unused params with underscore. In case, we can add
```suggestion
    "no-underscore-dangle": ["error", { "allowFunctionParams": true }],
```

See https://eslint.org/docs/rules/no-underscore-dangle#options

File: recommended.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    "no-console": "error",
+    "no-caller": "error",
+    "no-bitwise": "error",
+    eqeqeq: ["error", "smart"],
+    "max-classes-per-file": ["error", 1],
+    "guard-for-in": "error",
+    complexity: "error",
+    "arrow-body-style": "error",
+    "import/no-internal-modules": "off",
+    "import/order": "error",
```
Comment (balanza) [inline recommended.js:44]: I'm fine in having this, but I question if it's really useful

**Review by balanza — COMMENTED** [review]
LGTM apart minor comments

**Comment by balanza** [comment]
Great job guys! I'm loving this <3

File: recommended.js (hunk 9 above)
Comment (BurnedMarshal) [reply recommended.js:30]: It is, `"allowFunctionParams": true (default) allows dangling underscores in function parameter names`

Comment (balanza) [reply recommended.js:30]: So we want it or not? I think we want to be able to write
```ts
const fn = (_p) => {}
```

File: package.json [hunk 11]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "eslint-config",
```
Comment (balanza) [inline package.json]: Should we call it `@pagopa/eslint-config`?
```suggestion
  "name": "@pagopa/eslint-config",
```

File: recommended.js (hunk 9 above)
Comment (BurnedMarshal) [reply recommended.js:30]: Sorry, maybe my comment wasn't clear. We can write `const fn = (_p) => {}`, the option `allowFunctionParams` has `true` as default value.

File: package.json (hunk 11 above)
Comment (BurnedMarshal) [reply package.json]: You are right, is needed the scope into the package name.

File: package.json [hunk 12]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "eslint-config",
+  "name": "@pagopa/eslint-config",
```
Comment (balanza) [inline package.json:2]: <3

File: azure-pipelines.yml [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+    include:
+    - v*.*.*
+
+steps:
+  - task: Npm@1
+    inputs:
+      command: 'publish'
+      workingDir: '.'
+      publishEndpoint: 'NPM Registry'
+      verbose: true
```
Comment (balanza) [inline azure-pipelines.yml:12]: this won't work as per this issue https://github.com/microsoft/azure-pipelines-tasks/issues/14165
However we implemented a workaround here https://github.com/pagopa/io-functions-commons/pull/131

I suggest you to do it in another PR in which you can refactor the whole pipeline

**Review by balanza — APPROVED** [review]
